rl_studio/agents/autoparking/train_qlearn.py

```diff
@@ -0,0 +1,329 @@
+            "height"
+        ]
+        self.environment["width_image"] = params.agent["params"]["camera_params"][
+            "width"
+        ]
+        self.environment["center_image"] = params.agent["params"]["camera_params"][
+            "center_image"
+        ]
+
+        # Environment States
+        self.environment["state_space"] = params.agent["params"]["states"][
+            "state_space"
+        ]
+        self.environment["states"] = params.agent["params"]["states"][self.state_space]
+        # self.environment["x_row"] = params.agent["params"]["states"][self.state_space][0]
+
+        # Environment Actions
+        self.environment["action_space"] = params.environment["actions_set"]
+        self.environment["actions"] = params.environment["actions"]
+
+        # Environment Rewards
+        self.environment["reward_function"] = params.agent["params"]["rewards"][
+            "reward_function"
+        ]
+        self.environment["rewards"] = params.agent["params"]["rewards"][
+            self.reward_function
+        ]
+        self.environment["min_reward"] = params.agent["params"]["rewards"][
+            self.reward_function
+        ]["min_reward"]
+        self.environment["goal_reward"] = params.agent["params"]["rewards"][
+            self.reward_function
+        ]["goal_reward"]
+
+        # Environment Algorithm
+
+        #
+        self.environment["ROS_MASTER_URI"] = params.settings["params"]["ros_master_uri"]
+        self.environment["GAZEBO_MASTER_URI"] = params.settings["params"][
+            "gazebo_master_uri"
+        ]
+        self.environment["telemetry"] = params.settings["params"]["telemetry"]
+
+        print_messages(
+            "environment",
+            environment=self.environment,
+        )
+
+        # Env
+        self.env = gym.make(self.env_name, **self.environment)
+
+    def main(self):
+        os.makedirs(f"{self.outdir}", exist_ok=True)
+        start_time = datetime.now()
+        start_time_format = start_time.strftime("%Y%m%d_%H%M")
+
+        qlearn = QLearn(
+            actions=range(self.actions_size),
+            epsilon=self.epsilon,
+            alpha=self.alpha,
+            gamma=self.gamma,
+        )
+
+        ## Epsilon
+        epsilon_decay = self.epsilon / (self.total_episodes // 2)
+
+        # Checking state and actions
+        print_messages(
+            "In train_qlearn.py",
+            actions=self.actions,
+            # epsilon_decay=epsilon_decay,
+        )
+
+        ## -------------    START TRAINING --------------------
+        print(LETS_GO)
+        for episode in tqdm(
+            range(1, self.total_episodes + 1), ascii=True, unit="episodes"
+        ):
+            done = False
+            cumulated_reward = 0
+            step = 0
+            start_time_epoch = datetime.now()
+
+            observation = self.env.reset()
+            state = "".join(map(str, observation))
+
+            # ------- WHILE
```

Comment:
I would quit this comment. I think your code is clean enough to understand without this one

Reply:
ok, it is true, when coding for research, we don't realize in personal comments